File: Leetcode/Greedy/646_maximum_length_of_pair_chain.py
class Solution(object):

    def overlapped(self,interval1, interval2):
    	if interval1[1]>=interval2[0] and interval1[0]<=interval2[1]:
    		return True
    	else: 
    		return False

    def findLongestChain(self, pairs):
        """
        :type pairs: List[List[int]]
        :rtype: int
        """

        pairs.sort(key=lambda interval: interval[0])
        result, prev=0,0

        for i in xrange(1, len(pairs)):
        	if self.overlapped(pairs[i],pairs[prev]):
        		prev= prev if pairs[i][1] > pairs[prev][1] else i
        		result+=1

        		

        	else:
        		prev=i

        return len(pairs)-result
s=Solution()
print s.findLongestChain([[1,2], [2,3], [3,4]])

File: Leetcode/Dynamic_Programming/646_Maximum_Length_of_Pair_Chain.cpp
class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        
        if (pairs.size() == 0)
            return 0;
        
        auto compare=[](const vector<int> & a, const vector<int> b)
        {
            return a[1]<b[1];
        };
        sort(pairs.begin(),pairs.end(),compare);
        
        int down=INT_MIN;
        int n=0;
        
        for(int i=0;i<pairs.size();i++)
        {
            if(pairs[i][0]>down)
            {
                down=pairs[i][1];
                n++;
            }
        }
        return n;
    }
};